### ISSUE TITLE
Feature Request for LuaTable - clear()

### ISSUE THREAD

**Issue by Zamiell** [TypeScriptToLua/TypeScriptToLua#1051]
Recently, `has()` and `delete()` were added to LuaTable, which was a life saver. Is it possible to also add a `clear()` method?

The workaround right now is to blow away the existing table with a new table:

```ts
let myMap = new LuaTable<string, [number, number, number]();
// More code here
myMap = new LuaTable<string, [number, number, number](); // This is a pain to type
```

As you can see, it would be a lot easier if we could just type `.clear()`.
Furthermore, this workaround is annoying because you can't declare myMap as a const.

**Comment by tomblind** [comment]
You actually don't need the generic parameters in the re-assignment here. This works fine:
```ts
let myMap = new LuaTable<string, [number, number, number]>();
myMap = new LuaTable(); // Generic parameters are inferred here
```

**Comment by Zamiell** [comment]
That's true, but it still has the problems of:

1) `.clear()` is faster to type (and autocomplete) than ` = new LuaTable()`
2) You have to declare your LuaTable using `let`, which is dangerous.

**Comment by tomblind** [comment]
To be honest, the fact that this would allow you to reassign a const variable sounds like a bad idea.

Example:
```ts
const t = new LuaTable();
t.set("foo", "bar");
const t2 = t; // t2 and t are the same object
t.clear(); // t is now a new object
console.log(t2.get("foo")); // I would expect this to be nil, but it's "bar" because a new object was generated
```

**Comment by Zamiell** [comment]
Isn't that precisely how normal Maps work in JavaScript/TypeScript?

**Comment by tomblind** [comment]
No, Map.clear() removes everything from the map, but doesn't replace it.

[TS Playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBLC8MwFMDuMCyBDADgCgEoBuAKCgDoJko8AiAMxBFoBoZaAjLAJ1uJNCRYUAEwxEUUhWAAbZD0KlBEEHPIyQAczyjym6nUbMCxGAHozMAK5gAJsnoBLFLZJA)

**Comment by Zamiell** [comment]
Oh, I see the problem now.

**Comment by lolleko** [comment]
I think a clear() method would be nice, but like @tomblind said it should remove the entries from the table, not replace the table.

**Comment by Zamiell** [comment]
Looks like removing all entries of a table is an O(N) operation?
https://stackoverflow.com/questions/4880368/how-to-delete-all-elements-in-a-lua-table/4880638

**Comment by lolleko** [comment]
Yes, clearing would be O(n). But reassigning and letting the garbage collector do the cleanup is probably not much faster.
As always, whether this performance impact matters will depend on the use case, but in most cases O(n) shouldn't be a problem since the performed operation within the loop is extremely cheap.

**Comment by Zamiell** [comment]
Ok, so the feature is saved then. =)

**Comment by tomblind** [comment]
The core idea with LuaTable was to provide ways to directly access lua functionality so things could be done optimally when needed. Adding lib functions like an O(n) clear seems to get away from that purpose. I feel that once you starting needing those features, why not just use Map?

**Comment by lolleko** [comment]
I guess you are right, JS has not `Object.clear` method either.
I would be fine with adding a clear since it is not much work. But I would also be totally fine with not having one, for the reasons @tomblind outlined, and because it can easily be poly filled as needed with a simple reassign or a loop.

**Comment by Zamiell** [comment]
> why not just use Map?

I think a common use-case for TSTL is writing mods for games. In many video games, you can quit the game, close the program, and then come back and continue the game later on where you left off at. Thus, in the generic sense, any mod will have to save all of their state to disk.

TSTL maps use metatables in their implementation. This is a problem because they can't be converted to JSON and/or saved to disk cleanly without the end-user writing some kind of manual conversion layer (yuck).

A simple solution for this is for end-users to just use a LuaTable as a drop-in replacement for a map, since it has identical `get`, `set`, `has`, and `delete` methods. The only (major) thing missing is `clear` method.

**Comment by lolleko** [comment]
> TSTL maps use metatables in their implementation. This is a problem because they can't be converted to JSON and/or saved to disk cleanly without the end-user writing some kind of manual conversion layer (yuck).

I think serializing a Map is pretty simple `new Map(JSON.parse(JSON.stringify([...map.entries()])))`

> A simple solution for this is for end-users to just use a LuaTable as a drop-in replacement for a map, since it has identical get, set, has, and delete methods. The only (major) thing missing is clear method.

Map has many other functions, `entries()`, `keys()`, `values()`, `forEeach()`, so just adding clear wouldn't lead to identical interfaces.

But like I said I have no strong opinion on adding `clear` I'm fine with it either way.
